src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -1556,12 +1556,22 @@ public static SocketError AcceptAsync(Socket socket, SafeCloseSocket handle, Saf
 
         internal static SocketError DisconnectAsync(Socket socket, SafeCloseSocket handle, bool reuseSocket, DisconnectOverlappedAsyncResult asyncResult)
         {
-            throw new PlatformNotSupportedException(SR.net_sockets_disconnect_notsupported);
+            SocketError socketError = Disconnect(socket, handle, reuseSocket);
+            if (socketError == SocketError.Success)
+            {
+                asyncResult.CompletionCallback(SocketError.Success);
```

Comment:
What if it's not socketError == SocketError.Success?  When will the asyncResult be completed in that case?  Seems like in the case of an error, this would cause a hang, as the asyncresult will never be completed.